docs/initrd_generation.md

```diff
@@ -0,0 +1,138 @@
+as a separate file.
+
+When building a bootable image with `mkosi`, an initrd is generated as part of
+the build. By default, `SplitArtifacts=` includes the `initrd` option, so a copy
+of the final initrd is saved alongside other build artifacts.
+
+If you inspect this initrd file using common tools such as `lsinitrd` or
+`unmkinitramfs`, you may notice that kernel modules and firmware files do not
+appear in the output. To understand why, it is necessary to examine the internal
+structure of initrd files.
+
+# Structure of initrd Files
+
+Historically, an initrd file has consisted of multiple components. The first
+component, commonly referred to as the *early cpio*, is an optional, uncompressed
+cpio archive containing CPU microcode updates. The kernel attempts to locate and
+apply these updates very early in the boot process.
+
+The remainder of the initrd contains the early userspace, loadable kernel
+modules, and firmware files. In early Linux systems, these files were stored
+as a filesystem image, typically using ext2. This approach was later replaced
+by the use of an initramfs, which stores files in one or more cpio archives
+rather than in a filesystem image. Despite the terminology, what is commonly
+called an “initrd” today usually contains an initramfs rather than an ext2
+initrd image.
+
+From the beginning, the
+[kernel code](https://github.com/torvalds/Linux/blob/1da177e4c3f41524e886b7f1b8a0c1fc7321cac2/init/initramfs.c#L511)
+responsible for extracting the initramfs has
+[supported](https://www.kernel.org/doc/html/v6.18/driver-api/early-userspace/buffer-format.html)
+multiple concatenated cpio archives, each optionally compressed. These
+archives are unpacked sequentially to populate the in-memory `rootfs` used
+during early boot.
+
+`mkosi` uses this mechanism to split the initramfs into distinct parts:
+
+* **Default initrd**: contains only the early userspace, primarily `init` and its
+  supporting files. This archive is kernel-version independent.
+* **Kernel Module initrd (KMI)**: contains kernel modules and firmware files, which
+  are specific to a particular kernel version.
+
+For historical reasons, commonly used initrd inspection tools such as
+`lsinitrd`, `unmkinitramfs`, and even newer tools like `3cpio`, only process the
+early cpio and the first cpio archive in the initrd. Any additional concatenated
+archives are ignored.
+
+As a result, kernel modules included in the KMI do not appear when inspecting a
+typical `mkosi`-generated initrd with these tools.
+
+# Inspecting the Kernel Module initrd
+
+To work around the limitations of existing tools, `mkosi` provides the
+`SplitArtifacts=kernel-modules-initrd` option. When enabled, `mkosi` saves the
+KMI as a separate file, making it directly inspectable.
+
+The KMI is a cpio archive, optionally compressed, and can be examined using
+standard tools:
+
+```
+# Rebuild and generate a separate KMI file
+$ mkosi --split-artifacts=kernel-modules-initrd -f
```

Comment:
This is not really helpful for somebody trying to follow along. In an empty directory it doesn't produce the split output and in the mkosi repo it will be `mkosi.output/image.kernel-modules-initrd`.